system/core/document.php

```diff
@@ -568,11 +569,52 @@ public function dataFor($name, $default = null) {
 			return $default;
 		}
 
-		function validateRequiredField($name) {
-			if (!array_key_exists($name, $this->fields)) {
-				$this->errors[$name] = __('field-not-found');
-				return false;
+		/**
+		 * Is this form valid? Returns true when:
+		 *  - validate() was called
+		 *  - all of the fields for which validateXXXField
+		 *    methods were called were also valid.
+		 */
+		function isValid() {
+			return $this->is_validated && empty($this->errors);
+		}
+
+		/**
+		 * Validate this form against rules expressed in the
+		 * form itself. This must be called on a form before
+		 * processing it, otherwise `isValid()` will return
+		 * false.
+		 *
+		 * Currently, this validates:
+		 *  - Fields with the `required` attribute set.
+		 *  - Fields with the `type=email` set.
+		 *  - Editor fields for safe HTML
+		 */
+		function validate() {
+			foreach ($this->fields as $name => $elems) {
+				foreach ($elems as $elem) {
+					if ($elem->hasAttribute('required'))
+						$this->validateRequiredField($name);
+					if ($elem->tagName == 'input' && $elem->getAttribute('type') == 'email')
```

Comment:
If an email field is empty I think you would like the message 'required field missing', not 'incorrect email address' which seems to happen now that the first error gets overwritten by the second? Perhaps shift the order? Or make the errors per name into an array instead?

Reply:
Ah, no because the an empty email field is considered valid, the error won't be overwritten.

Reply:
Exactly.